src/ship/GunManager.h

```diff
@@ -0,0 +1,228 @@
+		float overheatThreshold = 1; // total amount of thermal energy(kJ) the gun can store while functioning
+
+		ProjectileType projectileType = PROJECTILE_BALLISTIC;
+		uint8_t numBarrels = 1;		 // total number of barrels on the model
+		bool staggerBarrels = false; // should we fire one barrel after another, or both at the same time?
+
+		ProjectileData projectile; // deprecated, to replace with RefCountedPtr<ProjectileDef>
+		std::string modelPath; // model to render this weapon with
+	};
+
+	// Information about a specific mounted weapon (serialized directly)
+	struct WeaponState {
+		WeaponMount *mount;					// Mounted hardpoint
+		uint8_t group = 0;					// Group this hardpoint belongs to
+		bool withinGimbalLimit = false;		// does the gun have a valid firing solution on the target?
+
+		uint8_t lastBarrel = 0;	 // the last barrel used to fire (for multi-barrel weapons)
+		float temperature = 0;	 // current gun temperature
+		double nextFireTime = 0; // time at which the gun will be ready to fire again
+		// TODO: integrate this with a whole-ship cooling system
+		// Currently uses GunManager::m_coolingBoost
+		// float coolerOverclock = 1.0; // any boost to gun coolers from ship equipment
+		// float powerOverclock = 1.0; // boost to overall firerate (and corresponding increase in heat generation)
+
+		vector3f currentLead;
+		vector3d currentLeadPos;
+
+		WeaponData data;
+		SceneGraph::Model *model; // gun model, currently unused
+	};
+
+	// Information about a specific mount that a weapon is attached to
+	// Currently only handles gimballed weapon mounts, but may support turrets in the future
+	struct WeaponMount {
+		StringName id;
+		SceneGraph::Tag *tag;     // Tag in the parent model that this weapon mount is attached to
+		vector2f gimbalLimitTan;  // tangent of gimbal limits
+		//SceneGraph::Model *model; // model to render for this weapon mount, if any
+		// TODO: enable/disable hardpoint based on ship configuration, i.e. landing/vtol/wings?
+	};
+
+	// Combines one or more weapons with a shared fire-control trigger and targeting information
+	struct GroupState {
+		WeaponIndexSet weapons;				// Whic weapons are assigned to this group?
+		const Body *target;					// The target for this group, if any
+		bool firing;						// Is the group currently firing
+		bool fireWithoutTargeting;			// Can the group fire without a target/target not in gimbal range?
+		ConnectionTicket onTargetDestroyed;	// Unlock the target once it's been destroyed
+	};
+
+	GunManager() = default;
+
+	void Init(ModelBody *b);
+
+	void SaveToJson(Json &jsonObj, Space *space);
+	void LoadFromJson(const Json &jsonObj, Space *space);
+
+	void StaticUpdate(float deltaTime);
+
+	// ==========================================
+
+	// Add a weapon mount to this gun manager.
+	// Returns false if a hardpoint already exists on this GunManager with the specified name.
+	bool AddWeaponMount(StringName id, StringName tagName, vector2f gimbalLimitDegrees);
```

Comment:
Nitpick: parameters should be const-ref's?  (ditto all other functions taking StringName parameters?)

(Going by JsonUtils.cpp using const-ref's for StringName types)

Reply:
In this case, the const-refs are to satisfy the function prototype of `nlohman::to_json`, used for implicit object<->JSON conversion. `StringName` itself is technically intended as a pass-by-value object, as it has small-string optimization to avoid memory allocation and copies only increment an internal refcount even for large strings.

...with that being said, I can (and might) convert these to const ref parameters as the refcount-copy operation does technically have a small fixed cost.